### DISCUSSION TITLE
VS Code Package?

### DISCUSSION THREAD

**Discussion by ttscoff** [ttscoff/bunch-docs#138] [Ideas]
There are now [Bunch IDE integrations](https://bunchapp.co/docs/integration/ide-packages/) for both Sublime Text and TextMate (thanks @kotfu). I'd _love_ to see one for VS Code, but I don't have the time/energy to learn how to write a package for an app I'm not currently using. I get why people like VS Code, though, and I would wager it's the most popular IDE out there right now, so Bunch support would be awesome.

If anyone wants to help out, you can see the code for both existing packages on GitHub. Between the two, all of the regexes you'd need to add syntax highlighting should already be written:

- [Bunch package for Sublime Text](https://github.com/ttscoff/Sublime-Bunch)
- [Bunch bundle for TextMate](https://github.com/kotfu/Bunch.tmbundle)

If you have the time/motivation to port to VS Code and need anything from me, just let me know!

**Comment by ttscoff** [comment]
Apparently it's really easy to use tmlanguage files in VS Code. Initial version of a VS Code extension: https://github.com/ttscoff/VSCode-Bunch

Anyone want to pitch in on expanding it? I'll gladly add maintainers to that repo.

**Reply by kotfu** [reply]
I'm happy to give it a go, I use VS Code a bit from time to time.

**Reply by ttscoff** [reply]
I keep wanting to like VS Code, but I can't get past the Electron inconveniences. But if you have a bit more experience with it than I do, I'd love to see what you come up with!

**Reply by ttscoff** [reply]
@kotfu do you want to be added as a maintainer to the one I started, or start one from scratch? The existing repo is just a `yo code` starter that imported your `.tmlanguage` file, so easy to start again if you prefer.

**Reply by kotfu** [reply]
I share your dislike of Electron apps, including VS Code, which is why I don't use it much. Just add me to the existing repo and I'll see what I can figure out.

**Reply by ttscoff** [reply]
Sent a collaborator invite, I think that gives you push access. Been a while, though, let me know if you need other permissions.

**Comment by claudinec** [comment]
First of all, thank you @ttscoff for Bunch! I'm a pretty new user and as someone who prefers to edit text, I'm finding Bunch is a powerful complement to Keyboard Maestro.

VSCode is my main text editor, so thanks Brett and @kotfu for working on this. I've briefly tested this extension and can confirm that syntax highlighting works.

I haven't developed VSCode extensions but could help with documentation if that would be useful.

![Screen Shot 2021-08-18 at 11 37 45 am](https://user-images.githubusercontent.com/74594/129824772-9b12147a-03c7-4807-9099-bf349d66040a.png)

![Screen Shot 2021-08-18 at 11 37 31 am](https://user-images.githubusercontent.com/74594/129824788-2010ba15-4b55-40d4-90b5-c222edffb9b4.png)

**Comment by ttscoff** [comment]
@kotfu I added new syntax for heredocs in the current beta. It will affect both the TM and VSC bundles, but doesn't break any existing highlighting. It affects [AppleScript and Shell Script lines](http://bunchapp.co/docs/bunch-files/scripts/shell-scripts/#heredoc), as well as [variable assignments](https://bunchapp.co/docs/bunch-files/variables/#heredoc).

There's also new support for [using snippet syntax](https://bunchapp.co/docs/bunch-files/variables/#snippet) on variable assignments and script lines.

Just in case you weren't busy enough :).

**Reply by kotfu** [reply]
If you used strict heredoc syntax then you couldn't have a display clause on the end. But since you get to pick the syntax, are display and/or delay clauses allowed after either markdown or heredoc embedded scripts, ie:
```
$ <<EOF
#!/bin/bash
ps aux | head -10
EOF (display) ~2
```
or
```
$ ```
ps aux | head -10
``` (display d:1 100%x300 left,top #333 #b0d17d a:95% l:d) ~4
```

Perhaps instead of "Are" the question is: Should display and delay clauses be allowed after either markdown or heredoc embedded scripts?

**Reply by ttscoff** [reply]
Yes, display and delay should definitely be an option for these. I'll add that next time I sit down to work on it. I don't think it will be a big deal, but I always guess wrong on these things.

**Reply by ttscoff** [reply]
Ok, display and delay both work with the heredoc syntax (as of beta 121). I've updated the Sublime package to correctly highlight with those trailing the closing fence/marker.

**Comment by kotfu** [comment]
Also, should you be able to use the embedded script syntax when assigning to a variable, like:
```
var3 = $ <<DUN
echo hi there
DUN
```

**Reply by ttscoff** [reply]
I'll still need to look into this. Might get a bit complicated.

**Comment by kotfu** [comment]
@ttscoff, do you have (or can you make) a png version of the bunch icon with a transparent background? The VSCode Marketplace lets you add an icon, but it has to be local to your repository (i.e. no linking to assets on the web). I used I used the one at https://cdn.bunchapp.co/images/bunchicon-350@2x.webp, but when it is displayed by VSCode, it renders black for the background instead of making it transparent.
<img width="384" alt="Screen Shot 2021-09-04 at 9 28 54 PM" src="https://user-images.githubusercontent.com/218536/132113941-04591cab-5428-4eac-a83c-3b703611b776.png">

If so, just commit it in https://github.com/ttscoff/VSCode-Bunch/tree/main/gfx. I'll get it hooked up properly.

Thx

**Comment by ttscoff** [comment]
There should be a PNG version at 
***@***.*** (and a 350px version 
if you remove the @2x). Let me know if that doesn't work, maybe you need 
a square version _without_ transparent bg?

**Reply by ttscoff** [reply]
Oh wow, I didn't realize it censored things that look like email addresses in emailed responses. In case you didn't figure it out, the png version is just the .webp substituted with .png.

**Reply by kotfu** [reply]
All the web assets aren't updated with the fancy new icon from 1.4.5.

i.e. `https://cdn.bunchapp.co/images/bunchicon-350@2x.webp` isn't the same picture as `https://cdn.bunchapp.co/images/bunchicon-350.png`.

**Comment by kotfu** [comment]
The Visual Studio Code Extension for Bunch is now available in the [Visual Studio Code Marketplace](https://marketplace.visualstudio.com/items?itemName=kotfu.bunch). When you open a bunch file in VS Code, it will prompt you to install the extension. To install manually:

- Launch Visual Studio Code
- Click the "Extensions" button in the sidebar
- Search for "Bunch" and click on the one with the icon showing the bunch of grapes
- Click Install

Updates will be installed automatically by Visual Studio Code.

**Reply by claudinec** [reply]
Brilliant, thank you!

**Comment by kotfu** [comment]
@ttscoff, you've got a section for the VS Code extension in integration/advanced-scripting/ide-packages.md that now needs to be updated. That section appears only in the main branch, not develop. Do you want a PR for these updates on the main branch or develop? I've also got a screenshot to include, but that can't go in the PR. How would you advise me to proceed?

**Comment by ttscoff** [comment]
I’ll need to pull the screenshot into the assets folder which isn’t in the repo, so make a pr on develop and include a link to the image in the comment. Thanks!

(I still need to get back to you on heredoc thoughts, but am enjoying a long weekend AFK.)

- Brett

**Comment by ttscoff** [comment]
Oh, and let me go fix the main/develop mismatch right now.

- Brett

**Comment by kotfu** [comment]
Version 1.1.0 of the VSCode extension has been published to the marketplace. VSCode will automatically update this for you. Version 1.1.0 of the extension contains support for the new syntaxes added in Bunch 1.4.5.

**Comment by ttscoff** [comment]
Cached on the CDN. You an either get it directly from the site by taking 
`cdn.` out of the url, or add a cache buster like:

	***@***.***

Any random number `.[BUSTER].png` will do it.

-Brett
